### ISSUE TITLE
Inefficient parsing when using jx and --args

### ISSUE THREAD

**Issue by btovar** [cooperative-computing-lab/cctools#2221]
The following takes less than a second to parse:

```json
# file: main_embedded.jx
{
    "define": {
        "DATA_FILE_LIST" : [ { "PATH" : "somepath." + n } for n in range(1500) ]
    },
    "rules": [
        {
            "command" : "echo hello >" + DATA["PATH"],
            "outputs" : [ DATA["PATH"] ]
        } for DATA in DATA_FILE_LIST
    ]
}
```

If instead we generate DATA_FILE_LIST to use in an args file as:

```json
#file: file_list_raw.jx
{
    "DATA_FILE_LIST": [ { "PATH" : "somepath." + n } for n in range(1500) ]
}

#file: main_external.jx
{
    "rules": [
        {
            "command" : "echo hello >" + DATA["PATH"],
            "outputs" : [ DATA["PATH"] ]
        } for DATA in DATA_FILE_LIST
    ]
}
```

$ jx2json file_list_raw.jx > file_list.jx
$ jx2json main_external --args file_list.jx

it takes about 40 s (which is much more than 1s when not using --args).

Any ideas @trshaffer, @dthain ?

**Comment by dthain** [comment]
Nothing immediately jumps to mind, but I might guess that some internal `eval` is either too lazy or too eager, resulting in n^2 behavior for the double-for.

I would be inclined to jus run both examples through gprof to see where time is being spent.  A different number of calls to `jx_eval` would be a good hint.

**Comment by btovar** [comment]
Many functions that evaluate jx do a jx_copy on the context.  The variables defined in --args end up as the context of the workflow.

For the examples above, using --args there are 197076363 calls to jx_copy. Without args, there are only 1613046. An increase of three orders of magnitude. 

Fixing this one will require some thought...

**Comment by dthain** [comment]
Actually, I don't see that many instances of copying the context.

There is one in the top-level `jx_eval` that doesn't make sense to me.
And then there is one in `jx_eval_comprehension` that is more understandable.

It seems to me there is a substantial difference between `jx_eval(jx_copy( big list ))` and `jx_copy(jx_eval(list comprehension))`

Does that make sense?

**Comment by btovar** [comment]
Yes, by commenting out the jx_copy/jx_delete in jx_eval parsing is done in 2.5s.

@trshaffer, do you see any problems with this?

**Comment by btovar** [comment]
solved by #2224
